sql/core/src/main/java/org/apache/spark/sql/execution/RecordBinaryComparator.java

```diff
@@ -38,32 +43,39 @@ public int compare(
     // check if stars align and we can get both offsets to be aligned
     if ((leftOff % 8) == (rightOff % 8)) {
       while ((leftOff + i) % 8 != 0 && i < leftLen) {
-        final int v1 = Platform.getByte(leftObj, leftOff + i) & 0xff;
-        final int v2 = Platform.getByte(rightObj, rightOff + i) & 0xff;
+        final int v1 = Platform.getByte(leftObj, leftOff + i);
+        final int v2 = Platform.getByte(rightObj, rightOff + i);
         if (v1 != v2) {
-          return v1 > v2 ? 1 : -1;
+          return (v1 & 0xff) > (v2 & 0xff) ? 1 : -1;
         }
         i += 1;
       }
     }
     // for architectures that support unaligned accesses, chew it up 8 bytes at a time
     if (Platform.unaligned() || (((leftOff + i) % 8 == 0) && ((rightOff + i) % 8 == 0))) {
       while (i <= leftLen - 8) {
-        final long v1 = Platform.getLong(leftObj, leftOff + i);
-        final long v2 = Platform.getLong(rightObj, rightOff + i);
+        long v1 = Platform.getLong(leftObj, leftOff + i);
+        long v2 = Platform.getLong(rightObj, rightOff + i);
         if (v1 != v2) {
-          return v1 > v2 ? 1 : -1;
+          if (LITTLE_ENDIAN) {
```

Comment:
for better perf, shall we move the if out of the while loop? e.g.
```
if (LITTLE_ENDIAN) {
  while ...
    return Long.compareUnsigned(Long.reverseBytes(v1), Long.reverseBytes(v2))
} else {
  while ...
    return Long.compareUnsigned(v1, v2);
}
```

Reply:
hmm maybe doesn't matter as it's a constant. cc @kiszk @rednaxelafx

Reply:
Probably won't matter much as this executes only once per comparison, when we know it's going to stop one way or the other.

Reply:
ah you are right. then nvm

Reply:
Just to note: `LITTLE_ENDIAN` is declared as `static final boolean` here, which allows it to be a JIT-compile-time constant, so during JIT compilation this condition check can be completely folded away. So don't worry about it.